File: src/main/kotlin/com/amitkurud/config/apidocs/SwaggerAutoConfig.java
/*
 * Copyright (c) 2020. Amit Kurud
 */

package com.amitkurud.config.apidocs;

import com.google.common.base.Predicates;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.*;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.util.StopWatch;
import org.springframework.util.StringUtils;
import org.springframework.web.servlet.DispatcherServlet;
import springfox.bean.validators.configuration.BeanValidatorPluginsConfiguration;
import springfox.documentation.builders.PathSelectors;
import springfox.documentation.schema.AlternateTypeRule;
import springfox.documentation.service.ApiInfo;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spring.web.plugins.Docket;
import springfox.documentation.swagger2.annotations.EnableSwagger2;

import javax.servlet.Servlet;
import java.nio.ByteBuffer;
import java.util.*;

import static springfox.documentation.builders.PathSelectors.regex;

@Configuration
@ConditionalOnWebApplication
@ConditionalOnClass({
        ApiInfo.class,
        BeanValidatorPluginsConfiguration.class,
        Servlet.class,
        DispatcherServlet.class,
        Docket.class
})
@EnableSwagger2
@Import(BeanValidatorPluginsConfiguration.class)
public class SwaggerAutoConfig {

    static final String STARTING_MESSAGE = "Starting Swagger";
    static final String STARTED_MESSAGE = "Started Swagger in {} ms";
    static final String MANAGEMENT_TITLE_SUFFIX = "Management API";
    static final String MANAGEMENT_GROUP_NAME = "management";
    static final String MANAGEMENT_DESCRIPTION = "Management endpoints documentation";

    private final Logger log = LoggerFactory.getLogger(SwaggerAutoConfig.class);

    @Bean
    @ConditionalOnMissingBean(name = "swaggerSpringfoxApiDocket")
    public Docket swaggerSpringfoxApiDocket(List<SwaggerCustomizer> swaggerCustomizers,
                                            ObjectProvider<AlternateTypeRule[]> alternateTypeRules) {
        log.debug(STARTING_MESSAGE);
        StopWatch watch = new StopWatch();
        watch.start();

        Docket docket = createDocket();

        // Apply all SwaggerCustomizers orderly.
        swaggerCustomizers.forEach(customizer -> customizer.customize(docket));

        // Add all AlternateTypeRules if available in spring bean factory.
        // Also you can add your rules in a customizer bean above.
        Optional.ofNullable(alternateTypeRules.getIfAvailable()).ifPresent(docket::alternateTypeRules);

        watch.stop();
        log.debug(STARTED_MESSAGE, watch.getTotalTimeMillis());
        return docket;
    }

    /**
     * JHipster Swagger Customizer
     *
     * @return the Swagger Customizer of JHipster
     */
    @Bean
    public ApiSwaggerCustomizer jHipsterSwaggerCustomizer() {
        return new ApiSwaggerCustomizer();
    }

    /**
     * Springfox configuration for the management endpoints (actuator) Swagger docs.
     *
     * @param appName               the application name
     * @param managementContextPath the path to access management endpoints
     * @return the Swagger Springfox configuration
     */
    @Bean
    @ConditionalOnClass(name = "org.springframework.boot.actuate.autoconfigure.web.server.ManagementServerProperties")
    @ConditionalOnProperty("management.endpoints.web.base-path")
    @ConditionalOnExpression("'${management.endpoints.web.base-path}'.length() > 0")
    @ConditionalOnMissingBean(name = "swaggerSpringfoxManagementDocket")
    public Docket swaggerSpringfoxManagementDocket(@Value("${spring.application.name:application}") String appName,
                                                   @Value("${management.endpoints.web.base-path}") String managementContextPath) {

        ApiInfo apiInfo = new ApiInfo(
                StringUtils.capitalize(appName) + " " + MANAGEMENT_TITLE_SUFFIX,
                MANAGEMENT_DESCRIPTION,
                "0.0.1-SNAPSHOT",
                "",
                ApiInfo.DEFAULT_CONTACT,
                "",
                "",
                new ArrayList<>()
        );

        return createDocket()
                .apiInfo(apiInfo)
                .useDefaultResponseMessages(true)
                .groupName(MANAGEMENT_GROUP_NAME)
                .forCodeGeneration(true)
                .directModelSubstitute(ByteBuffer.class, String.class)
                .genericModelSubstitutes(ResponseEntity.class)
                .ignoredParameterTypes(Pageable.class)
                .select()
                .paths(regex(managementContextPath + ".*"))
                .paths(Predicates.not(PathSelectors.regex("/error.*")))
                .build();
    }

    protected Docket createDocket() {
        return new Docket(DocumentationType.SWAGGER_2);
    }

}

File: src/main/kotlin/com/amitkurud/config/apidocs/SwaggerAutoConfig.kt
/*
 * Copyright (c) 2020. Amit Kurud
 */
package com.amitkurud.config.apidocs

import com.google.common.base.Predicates
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
import org.springframework.boot.info.BuildProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.DispatcherServlet
import springfox.bean.validators.configuration.BeanValidatorPluginsConfiguration
import springfox.documentation.builders.PathSelectors
import springfox.documentation.builders.RequestHandlerSelectors
import springfox.documentation.service.ApiInfo
import springfox.documentation.spi.DocumentationType
import springfox.documentation.spring.web.plugins.Docket
import springfox.documentation.swagger2.annotations.EnableSwagger2
import java.nio.ByteBuffer
import java.util.*
import javax.servlet.Servlet

@Configuration
@ConditionalOnWebApplication
@ConditionalOnClass(
    ApiInfo::class,
    BeanValidatorPluginsConfiguration::class,
    Servlet::class,
    DispatcherServlet::class,
    Docket::class
)
@EnableSwagger2
@Import(BeanValidatorPluginsConfiguration::class)
class SwaggerAutoConfig {
    private val log = LoggerFactory.getLogger(SwaggerAutoConfig::class.java)

    @Autowired
    private val buildProperties: BuildProperties? = null

    @Bean
    @ConditionalOnMissingBean(name = ["swaggerSpringfoxManagementDocket"])
    fun swaggerSpringfoxManagementDocket(
        @Value("\${spring.application.name:application}") appName: String?,
        swaggerCustomizers: List<SwaggerCustomizer?>?
    ): Docket {
        val apiInfo = ApiInfo(
            StringUtils.capitalize(appName) + " " + MANAGEMENT_TITLE_SUFFIX,
            mavenBuildInfo,
            "0.0.1-SNAPSHOT",
            "",
            ApiInfo.DEFAULT_CONTACT,
            "",
            "",
            ArrayList()
        )
        return createDocket().apiInfo(apiInfo)
    }

    private val mavenBuildInfo: String
        private get() {
            if (buildProperties == null) {
                return ""
            }
            var mavenBuildInfo = ""
            if (buildProperties.version != null) {
                mavenBuildInfo += "Version - " + buildProperties.version + ";"
            }
            if (buildProperties.time != null) {
                mavenBuildInfo += " Last build time - " + buildProperties.time.toString() + ";"
            }
            return mavenBuildInfo
        }

    protected fun createDocket(): Docket {
        return Docket(DocumentationType.SWAGGER_2)
            .useDefaultResponseMessages(true)
            .forCodeGeneration(true)
            .directModelSubstitute(ByteBuffer::class.java, String::class.java)
            .genericModelSubstitutes(ResponseEntity::class.java)
            .ignoredParameterTypes(Pageable::class.java)
            .select()
            .apis(RequestHandlerSelectors.withClassAnnotation(RestController::class.java))
            .paths(PathSelectors.any())
            .paths(Predicates.not(PathSelectors.regex("/error/*")))
            .build()
    }

    companion object {
        const val STARTING_MESSAGE = "Starting Swagger"
        const val STARTED_MESSAGE = "Started Swagger in {} ms"
        const val MANAGEMENT_TITLE_SUFFIX = "Management API"
        const val MANAGEMENT_GROUP_NAME = "management"
        const val MANAGEMENT_DESCRIPTION = "Management endpoints documentation"
    }
}